infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/Orchestrator.java

```diff
@@ -107,41 +111,49 @@ void query(CorfuPayloadMsg<OrchestratorMsg> msg, ChannelHandlerContext ctx, ISer
      *
      * Create and start a workflow on this orchestrator, if there is
      * an existing workflow that is executing on the same endpoint,
-     * then just return the corresponding workflow id. Dispatch is the only
-     * place where workflows are created based on reading activeWorkflows
-     * and therefore needs to be synchronized to prevent launching multiple
-     * workflows for the same endpoint concurrently.
+     * the orchestrator will notify the client that endpoint is busy.
+     * Dispatch is the only place where workflows are created based on
+     * reading activeWorkflows and therefore access needs to be synchronized
+     * to prevent launching multiple workflows for the same endpoint concurrently.
      *
      * @param msg corfu message containing the create workflow request
      * @param ctx netty ChannelHandlerContext
      * @param r   server router
      */
-    synchronized void dispatch(CorfuPayloadMsg<OrchestratorMsg> msg,
-                               ChannelHandlerContext ctx, IServerRouter r) {
+    void dispatch(CorfuPayloadMsg<OrchestratorMsg> msg,
+                  ChannelHandlerContext ctx, IServerRouter r) {
         CreateRequest req = (CreateRequest) msg.getPayload().getRequest();
 
-        UUID id = activeWorkflows.inverse().get(req.getEndpoint());
-        if (id != null) {
-            // A workflow is already executing for this endpoint, return
-            // existing workflow id.
-            OrchestratorResponse resp = new OrchestratorResponse(
-                    new CreateWorkflowResponse(id));
-            r.sendResponse(ctx, msg, CorfuMsgType.ORCHESTRATOR_RESPONSE
-                    .payloadMsg(resp));
-            return;
-        } else {
-            // Create a new workflow for this endpoint and return a new workflow id
-            IWorkflow workflow = req.getWorkflow();
-            activeWorkflows.put(workflow.getId(), req.getEndpoint());
-
-            CompletableFuture.runAsync(() -> {
-                run(workflow);
-            });
+        synchronized (activeWorkflows) {
```

Comment:
It's not really clear why only this block of code is synchronized.

Reply:
From the methods description
"if there is
     * an existing workflow that is executing on the same endpoint,
     * the orchestrator will notify the client that endpoint is busy."

Reply:
I see that, but why do you need the synchronization on activeworkflows, it doesn't seem to achieve what you want since there could be a race between workflow creation and this check?